File: libs/services/src/os/DropBoxManager.cpp
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#define LOG_TAG "DropBoxManager"

#include <android/os/DropBoxManager.h>

#include <android-base/unique_fd.h>
#include <binder/IServiceManager.h>
#include <binder/ParcelFileDescriptor.h>
#include <com/android/internal/os/IDropBoxManagerService.h>
#include <cutils/log.h>

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

namespace android {
namespace os {

using namespace ::com::android::internal::os;

DropBoxManager::DropBoxManager()
{
}

DropBoxManager::~DropBoxManager()
{
}

Status
DropBoxManager::addText(const String16& tag, const string& text)
{
    return addData(tag, reinterpret_cast<uint8_t const*>(text.c_str()), text.size(), IS_TEXT);
}

Status
DropBoxManager::addData(const String16& tag, uint8_t const* data,
        size_t size, int flags)
{
    sp<IDropBoxManagerService> service = interface_cast<IDropBoxManagerService>(
        defaultServiceManager()->getService(android::String16("dropbox")));
    if (service == NULL) {
        return Status::fromExceptionCode(Status::EX_NULL_POINTER, "can't find dropbox service");
    }
    ALOGD("About to call service->add()");
    vector<uint8_t> dataArg;
    dataArg.assign(data, data + size);
    Status status = service->addData(tag, dataArg, flags);
    ALOGD("service->add returned %s", status.toString8().c_str());
    return status;
}

Status
DropBoxManager::addFile(const String16& tag, const string& filename, int flags)
{
    int fd = open(filename.c_str(), O_RDONLY);
    if (fd == -1) {
        string message("addFile can't open file: ");
        message += filename;
        ALOGW("DropboxManager: %s", message.c_str());
        return Status::fromExceptionCode(Status::EX_ILLEGAL_STATE, message.c_str());
    }
    return addFile(tag, fd, flags);
}

Status
DropBoxManager::addFile(const String16& tag, int fd, int flags)
{
    if (fd == -1) {
        string message("invalid fd (-1) passed to to addFile");
        ALOGW("DropboxManager: %s", message.c_str());
        return Status::fromExceptionCode(Status::EX_ILLEGAL_STATE, message.c_str());
    }
    sp<IDropBoxManagerService> service = interface_cast<IDropBoxManagerService>(
        defaultServiceManager()->getService(android::String16("dropbox")));
    if (service == NULL) {
        return Status::fromExceptionCode(Status::EX_NULL_POINTER, "can't find dropbox service");
    }
    ALOGD("About to call service->add()");
    android::base::unique_fd uniqueFd(fd);
    android::os::ParcelFileDescriptor parcelFd(std::move(uniqueFd));
    Status status = service->addFile(tag, parcelFd, flags);
    ALOGD("service->add returned %s", status.toString8().c_str());
    return status;
}

}} // namespace android::os

File: libs/services/src/dropboxmanager.rs
// Copyright (C) 2025 The Android Open Source Project
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Rust interface to the dropbox service.
use anyhow::Result;
use binder::{wait_for_interface, Strong};
use dropboxmanager_aidl::aidl::com::android::internal::os::IDropBoxManagerService::IDropBoxManagerService;

const INTERFACE_NAME: &str = "dropbox";

/// Interface to the DropBox system service.
pub struct DropBoxManager {
    binder: Strong<dyn IDropBoxManagerService>,
}

impl DropBoxManager {
    /// Acquires the underlying binder interface.
    pub fn new() -> Result<Self> {
        Ok(Self { binder: wait_for_interface(INTERFACE_NAME)? })
    }

    /// Creates a dropbox entry with the supplied tag. The supplied text is passed as bytes to create the file contents.
    pub fn add_text(&self, tag: &str, text: &str) -> Result<()> {
        self.binder.addData(tag, text.as_bytes(), 2 /* DropBoxManager.java IS_TEXT */)?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use std::path::PathBuf;

    const DROPBOX_PATH: &str = "/data/system/dropbox";
    const TAG: &str = "foo";
    const CONTENT: &str = "bar\nbaz\n";

    #[test]
    fn add_text() {
        let _ = find_dropbox_files(true).unwrap();
        let manager = DropBoxManager::new().unwrap();
        manager.add_text(TAG, CONTENT).unwrap();
        let path_buf = find_dropbox_files(false).unwrap().unwrap();
        let content = fs::read_to_string(path_buf.as_path()).unwrap();
        assert_eq!(content, CONTENT);
    }

    fn find_dropbox_files(delete_them: bool) -> Result<Option<PathBuf>> {
        let mut found = None;
        for entry in fs::read_dir(DROPBOX_PATH)? {
            let entry = entry?;
            let path_buf = entry.path();
            if !path_buf.is_file() {
                continue;
            }
            let Some(filename) = path_buf.file_name() else {
                continue;
            };
            let filename = filename.to_string_lossy();
            if !filename.starts_with(TAG) {
                continue;
            }

            found = Some(path_buf.clone());
            if delete_them {
                fs::remove_file(&path_buf)?;
            } else {
                break;
            }
        }
        Ok(found)
    }
}
